File: backend/mcp_servers/mysql_server.js
#!/usr/bin/env node

/**
 * MySQL MCP Server
 *
 * MCP 서버로 MySQL 데이터베이스에 연결하고 쿼리를 실행합니다.
 */

import { Server } from '@modelcontextprotocol/sdk/server/index.js';
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
import {
  CallToolRequestSchema,
  ListToolsRequestSchema,
} from '@modelcontextprotocol/sdk/types.js';
import mysql from 'mysql2/promise';
import dotenv from 'dotenv';

dotenv.config();

// MySQL 연결 풀 생성
const pool = mysql.createPool({
  host: process.env.MYSQL_HOST || 'localhost',
  port: parseInt(process.env.MYSQL_PORT || '3306'),
  database: process.env.MYSQL_DB,
  user: process.env.MYSQL_USER,
  password: process.env.MYSQL_PASSWORD,
  waitForConnections: true,
  connectionLimit: 10,
  queueLimit: 0,
});

// MCP 서버 생성
const server = new Server(
  {
    name: 'mysql-mcp-server',
    version: '0.1.0',
  },
  {
    capabilities: {
      tools: {},
    },
  }
);

// 도구 목록 제공
server.setRequestHandler(ListToolsRequestSchema, async () => {
  return {
    tools: [
      {
        name: 'query',
        description: 'Execute a read-only SQL query on MySQL database',
        inputSchema: {
          type: 'object',
          properties: {
            sql: {
              type: 'string',
              description: 'SQL query to execute (SELECT only)',
            },
          },
          required: ['sql'],
        },
      },
      {
        name: 'list_tables',
        description: 'List all tables in the current MySQL database',
        inputSchema: {
          type: 'object',
          properties: {},
        },
      },
      {
        name: 'describe_table',
        description: 'Get schema information for a specific table',
        inputSchema: {
          type: 'object',
          properties: {
            table_name: {
              type: 'string',
              description: 'Name of the table to describe',
            },
          },
          required: ['table_name'],
        },
      },
    ],
  };
});

// 도구 실행 핸들러
server.setRequestHandler(CallToolRequestSchema, async (request) => {
  const { name, arguments: args } = request.params;

  try {
    switch (name) {
      case 'query': {
        const { sql } = args;

        // 보안: SELECT 쿼리만 허용
        if (!sql.trim().toLowerCase().startsWith('select')) {
          throw new Error('Only SELECT queries are allowed');
        }

        const [rows] = await pool.query(sql);
        return {
          content: [
            {
              type: 'text',
              text: JSON.stringify({
                rows: rows,
                rowCount: Array.isArray(rows) ? rows.length : 0,
              }, null, 2),
            },
          ],
        };
      }

      case 'list_tables': {
        const [rows] = await pool.query('SHOW TABLES');
        const tableNames = rows.map(row => Object.values(row)[0]);
        return {
          content: [
            {
              type: 'text',
              text: JSON.stringify(tableNames, null, 2),
            },
          ],
        };
      }

      case 'describe_table': {
        const { table_name } = args;
        const [rows] = await pool.query(`DESCRIBE ${mysql.escapeId(table_name)}`);
        return {
          content: [
            {
              type: 'text',
              text: JSON.stringify(rows, null, 2),
            },
          ],
        };
      }

      default:
        throw new Error(`Unknown tool: ${name}`);
    }
  } catch (error) {
    return {
      content: [
        {
          type: 'text',
          text: `Error: ${error.message}`,
        },
      ],
      isError: true,
    };
  }
});

// 서버 시작
async function main() {
  const transport = new StdioServerTransport();
  await server.connect(transport);
  console.error('MySQL MCP Server running on stdio');
}

main().catch((error) => {
  console.error('Server error:', error);
  process.exit(1);
});

File: backend/mcp_servers/mysql_server.py
#!/usr/bin/env python3
"""
MySQL MCP Server

MCP 서버로 MySQL 데이터베이스에 연결하고 쿼리를 실행합니다.
"""
import json
import os
from mcp.server.fastmcp import FastMCP
import aiomysql
from dotenv import load_dotenv

from .lifecycle import setup_signal_handlers, register_cleanup, log_startup

load_dotenv()

mcp = FastMCP("mysql-mcp-server")
_pool = None


async def get_pool():
    """Lazy connection pool initialization for MySQL."""
    global _pool
    if _pool is None:
        _pool = await aiomysql.create_pool(
            host=os.getenv("MYSQL_HOST", "localhost"),
            port=int(os.getenv("MYSQL_PORT", "3306")),
            user=os.getenv("MYSQL_USER"),
            password=os.getenv("MYSQL_PASSWORD"),
            db=os.getenv("MYSQL_DB"),
            minsize=2,
            maxsize=10,
            autocommit=True,
            charset='utf8mb4'
        )
        register_cleanup(cleanup_pool)
    return _pool


async def cleanup_pool():
    """Close connection pool on shutdown."""
    global _pool
    if _pool:
        _pool.close()
        await _pool.wait_closed()
        _pool = None


@mcp.tool()
async def query(sql: str) -> str:
    """Execute a read-only SQL query on MySQL database.

    Args:
        sql: The SQL query to execute (SELECT only)

    Returns:
        JSON string with rows array and rowCount
    """
    # Security: SELECT only
    if not sql.strip().lower().startswith('select'):
        raise ValueError("Only SELECT queries are allowed")

    pool = await get_pool()
    async with pool.acquire() as conn:
        async with conn.cursor(aiomysql.DictCursor) as cursor:
            await cursor.execute(sql)
            rows = await cursor.fetchall()
            return json.dumps({
                "rows": rows,
                "rowCount": len(rows)
            }, indent=2, default=str)  # default=str handles datetime, Decimal, etc.


@mcp.tool()
async def list_tables() -> str:
    """List all tables in the current MySQL database.

    Returns:
        JSON array of table names
    """
    pool = await get_pool()
    async with pool.acquire() as conn:
        async with conn.cursor() as cursor:
            await cursor.execute("SHOW TABLES")
            rows = await cursor.fetchall()
            # SHOW TABLES returns tuples like ('table_name',)
            table_names = [row[0] for row in rows]
            return json.dumps(table_names, indent=2)


@mcp.tool()
async def describe_table(table_name: str) -> str:
    """Get schema information for a specific table.

    NOTE: MySQL DESCRIBE returns different columns than PostgreSQL information_schema:
    - Field: column name
    - Type: data type with size (e.g., 'varchar(255)')
    - Null: 'YES' or 'NO'
    - Key: 'PRI', 'UNI', 'MUL', or ''
    - Default: default value or None
    - Extra: additional info (e.g., 'auto_increment')

    Args:
        table_name: Name of the table to describe

    Returns:
        JSON array with column information (Field, Type, Null, Key, Default, Extra)
    """
    pool = await get_pool()
    async with pool.acquire() as conn:
        async with conn.cursor() as cursor:
            # Security: Validate table_name against information_schema to prevent SQL injection
            # Though we use backticks, direct validation is safer
            await cursor.execute(
                "SELECT table_name FROM information_schema.tables WHERE table_schema = %s AND table_name = %s",
                (os.getenv("MYSQL_DB"), table_name)
            )
            exists = await cursor.fetchone()
            if not exists:
                raise ValueError(f"Table '{table_name}' not found or access denied")

            # Use backticks for table name to handle reserved words
            await cursor.execute(f"DESCRIBE `{table_name}`")
            rows = await cursor.fetchall()
            # DESCRIBE returns: (Field, Type, Null, Key, Default, Extra)
            columns = ['Field', 'Type', 'Null', 'Key', 'Default', 'Extra']
            result = [dict(zip(columns, row)) for row in rows]
            return json.dumps(result, indent=2, default=str)


if __name__ == "__main__":
    setup_signal_handlers()
    log_startup("MySQL MCP Server")
    mcp.run(transport="stdio")
